README.md
# Robert-Koch Institut COVID numbers parser

[![License](https://img.shields.io/github/license/thebino/rki-covid-parser)](./LICENSE.md)
[![Tests](https://github.com/thebino/rki-covid-parser/actions/workflows/testing.yaml/badge.svg)](https://github.com/thebino/rki-covid-parser/actions/workflows/testing.yaml)
[![codecov](https://codecov.io/gh/thebino/rki-covid-parser/branch/development/graph/badge.svg?token=9NUUAMXVP4)](https://codecov.io/gh/thebino/rki-covid-parser)

Python parser for the latest covid numbers from German RKI (Robert-Koch Institut).

## Installation
```pip install rki-covid-parser```

## Usage
Initialize the parser with an `aiohttp.session` and load the latest data.
```
parser = RkiCovidParser(session)        
await parser.load_data()


for district in parser.districts:
    # work with districts
    pass

for state in parser.states:
    # work with states
    pass

# work with the country
parser.country.cases
```

finished loading data, the parser contains a dictionary of districts, each with these properties:

|Property    |Type|Description|
|:-----------|:---|:------------|
|`id`| string | Unique district identifier |
|`name`| string | Name of the  district |
|`county`| string | County of the district |
|`state`| string | State of the district |
|`population`| integer | Population |
|`cases`| integer | Active cases |
|`deaths`| integer | Currently tracked deaths |
|`casesPerWeek`| integer | Cases per week |
|`deathsPerWeek`| integer | Deaths per week |
|`recovered`| integer | Recovered cases |
|`weekIncidence`| float | Week incidence |
|`casesPer100k`| float | Cases per 100k population |
|`newCases`| integer | New cases since last day |
|`newDeaths`| integer | New deaths since last day |
|`newRecovered`| integer | New recovered since last day |
|`last_update`| string | Timestamp of the last update |
|`hospitalizationCasesMerged` | float | Hospitalization cases for people in germany summarized  |
|`hospitalizationIncidenceMerged` | float | Hospitalization incidence for people in germany summarized |
|`hospitalizationCasesBaby` | float | Hospitalization cases for people between 00-04 |
|`hospitalizationIncidenceBaby` | float | Hospitalization incidence for people between 00-04 |
|`hospitalizationCasesChildren` | float | Hospitalization cases for people between 05-14 |
|`hospitalizationIncidenceChildren` | float | Hospitalization incidence for people between 05-14 |
|`hospitalizationCasesTeen` | float | Hospitalization cases for people between 15-34 |
|`hospitalizationIncidenceTeen` | float | Hospitalization incidence for people between 15-34 |
|`hospitalizationCasesGrown` | float | Hospitalization cases for people between 35-59 |
|`hospitalizationIncidenceGrown` | float | Hospitalization incidence for people between 35-59 |
|`hospitalizationCasesSenior` | float | Hospitalization cases for people between 60-79 |
|`hospitalizationIncidenceSenior` | float | Hospitalization incidence for people between 60-79 |
|`hospitalizationCasesOld` | float | Hospitalization cases for people older than 80 |
|`hospitalizationIncidenceOld` | float | Hospitalization incidence for people older than 80 |


## Contribution
See [Contribution](https://github.com/thebino/rki-covid-parser/blob/development/CONTRIBUTING.md) for details.


CHANGELOG.md
## [1.3.3] - 2022-02-21
### Fixed
- Parsing of `IdLandkreis` from changed datasource

## [1.3.2] - 2022-02-20
### Fixed
- Change datasource for newCases, newDeaths

## [1.3.1] - 2021-11-28
### Fixed
- Skip unavailable values

## [1.3.0] - 2021-11-25
### Added
- Added hospitalization numbers

### Fixed
- Skip empty feature attributes

## [1.2.4] - 2021-09-13
### Fixed
- Stop parsing without exception, when data source is not available

## [1.2.3] - 2021-09-11
### Changed
- version bump

## [1.2.2] - 2021-09-10
### Fixed
- Skip non mapped states for vaccination numbers

## [1.2.1] - 2021-07-01
### Fixed
- Reset state values before merging

## [1.2.0] - 2021-04-28
### Changed
- Simplified state parsing
- Roundet incidence to only 1 decimal

## [1.1.0] - 2021-04-27
### Added
- Vaccination data parsing

## [1.0.0] - 2021-04-18
### Added
- Parsing districts, states and country "Deutschland"
- Arcgis endpoint availibility check workflow


tests/conftest.py
import pytest
import json
import datetime

@pytest.fixture()
def response_districts():
    with open("tests/fixtures/response_districts.json", "r") as mock_response:
        return json.load(mock_response)

@pytest.fixture()
def no_districts():
    with open("tests/fixtures/no_districts.json", "r") as mock_response:
        return json.load(mock_response)

@pytest.fixture()
def no_features():
    with open("tests/fixtures/no_features.json", "r") as mock_response:
        return json.load(mock_response)

@pytest.fixture()
def response_recovered():
    with open("tests/fixtures/response_recovered.json", "r") as mock_response:
        return json.load(mock_response)

@pytest.fixture()
def response_new_cases():
    with open("tests/fixtures/response_new_cases.json", "r") as mock_response:
        return json.load(mock_response)

@pytest.fixture()
def response_new_recovered():
    with open("tests/fixtures/response_new_recovered.json", "r") as mock_response:
        return json.load(mock_response)

@pytest.fixture()
def response_new_death():
    with open("tests/fixtures/response_new_death.json", "r") as mock_response:
        return json.load(mock_response)

@pytest.fixture()
def vaccinations_by_state():
    with open("tests/fixtures/germany_vaccinations_by_state.tsv", mode='r+b') as mock_response:
        return mock_response.read()

@pytest.fixture()
def hospitalization_numbers():
    # replace latest test-date with today
    text = open("tests/fixtures/hospitalization.csv", "r")
    text = ''.join([i for i in text]).replace("2021-11-24", str(datetime.date.today()))
    x = open("tests/fixtures/hospitalization.csv","w")
    x.writelines(text)
    x.close()

    with open("tests/fixtures/hospitalization.csv", mode='r+b') as mock_response:
        return mock_response.read()


tests/test_states.py
import asyncio
import json
import os

import aiohttp
import pytest

from rki_covid_parser.parser import RkiCovidParser
from rki_covid_parser.model.state import State

async def test_states_response():
    """Test the service response for content."""
    async with aiohttp.ClientSession() as session:
        parser = RkiCovidParser(session)
        await parser.load_data()

        assert len(parser.states) == 16
        for state in parser.states:
            validate_state(parser.states[state])

async def test_multiple_load_data_calls():
    """Test the service response for content."""
    async with aiohttp.ClientSession() as session:
        parser = RkiCovidParser(session)
        await parser.load_data()
        await parser.load_data()

        assert len(parser.states) == 16
        for state in parser.states:
            validate_state(parser.states[state])

def validate_state(state: State):
    # validate numbers against RKI Dashboard (https://experience.arcgis.com/experience/478220a4c454480e823b17327b2bf1d4/)
    if state.name == 'Bayern':
        assert state.cases == 653797
        assert state.weekIncidence == 14.4
    if state.name == 'Baden-Württemberg':
        assert state.cases == 505824
        assert state.weekIncidence == 14.8
    if state.name == 'Niedersachsen':
        assert state.cases == 265191
        assert state.weekIncidence == 16.8
    pass



tests/test_parsing.py
import aiohttp
from aioresponses import aioresponses

from typing import Dict

from rki_covid_parser.const import DISTRICTS_URL, DISTRICTS_URL_RECOVERED, DISTRICTS_URL_NEW_CASES, DISTRICTS_URL_NEW_RECOVERED, DISTRICTS_URL_NEW_DEATHS, HOSPITALIZATION_URL, VACCINATIONS_URL
from rki_covid_parser.parser import RkiCovidParser
from rki_covid_parser.model.district import District
from rki_covid_parser.model.state import State
from rki_covid_parser.model.country import Country

async def test_parsing_with_fake_data(response_districts, response_recovered, response_new_cases, response_new_recovered, response_new_death, vaccinations_by_state, hospitalization_numbers):
    """Test parsing with fake data loaded from fixture response files."""
    async with aiohttp.ClientSession() as session:
        with aioresponses() as mocked:
            mocked.get(DISTRICTS_URL, status=200, payload=response_districts)
            mocked.get(DISTRICTS_URL_RECOVERED, status=200, payload=response_recovered)
            mocked.get(DISTRICTS_URL_NEW_CASES, status=200, payload=response_new_cases)
            mocked.get(DISTRICTS_URL_NEW_RECOVERED, status=200, payload=response_new_recovered)
            mocked.get(DISTRICTS_URL_NEW_DEATHS, status=200, payload=response_new_death)
            mocked.get(VACCINATIONS_URL, status=200, body=vaccinations_by_state)
            mocked.get(HOSPITALIZATION_URL, status=200, body=hospitalization_numbers)

            parser = RkiCovidParser(session)
            await parser.load_data()

            assert len(parser.districts) == 412
            for district in parser.districts:
                _validate_district(parser.districts[district])

            for state in parser.states:
                _validate_state(parser.states[state])

            _validate_district_numbers(parser.districts)
            _validate_state_numbers(parser.states)
            _validate_country_numbers(parser.country)
            _validate_vaccination_numbers(parser.states)
            _validate_hospitalization_numbers(parser.states)

async def test_parsing_without_data(no_districts, response_recovered, response_new_cases, response_new_recovered, response_new_death, vaccinations_by_state, hospitalization_numbers):
    """Test parsing with fake data loaded from fixture response files."""
    async with aiohttp.ClientSession() as session:
        with aioresponses() as mocked:
            mocked.get(DISTRICTS_URL, status=200, payload=no_districts)
            mocked.get(DISTRICTS_URL_RECOVERED, status=200, payload=response_recovered)
            mocked.get(DISTRICTS_URL_NEW_CASES, status=200, payload=response_new_cases)
            mocked.get(DISTRICTS_URL_NEW_RECOVERED, status=200, payload=response_new_recovered)
            mocked.get(DISTRICTS_URL_NEW_DEATHS, status=200, payload=response_new_death)
            mocked.get(VACCINATIONS_URL, status=200, body=vaccinations_by_state)
            mocked.get(HOSPITALIZATION_URL, status=200, body=hospitalization_numbers)

            parser = RkiCovidParser(session)
            await parser.load_data()

            assert len(parser.districts) == 0


async def test_parsing_without_features(no_features, response_recovered, response_new_cases, response_new_recovered, response_new_death, vaccinations_by_state, hospitalization_numbers):
    """Test parsing with fake data loaded from fixture response files."""
    async with aiohttp.ClientSession() as session:
        with aioresponses() as mocked:
            mocked.get(DISTRICTS_URL, status=200, payload=no_features)
            mocked.get(DISTRICTS_URL_RECOVERED, status=200, payload=response_recovered)
            mocked.get(DISTRICTS_URL_NEW_CASES, status=200, payload=response_new_cases)
            mocked.get(DISTRICTS_URL_NEW_RECOVERED, status=200, payload=response_new_recovered)
            mocked.get(DISTRICTS_URL_NEW_DEATHS, status=200, payload=response_new_death)
            mocked.get(VACCINATIONS_URL, status=200, body=vaccinations_by_state)
            mocked.get(HOSPITALIZATION_URL, status=200, body=hospitalization_numbers)

            parser = RkiCovidParser(session)
            await parser.load_data()

            assert len(parser.districts) == 0


async def test_parsing_with_multiple_load_data_calls(response_districts, response_recovered, response_new_cases, response_new_recovered, response_new_death, vaccinations_by_state, hospitalization_numbers):
    """Test if parsing returns valid numbers if called multiple times."""
    async with aiohttp.ClientSession() as session:
        with aioresponses() as mocked:
            mocked.get(DISTRICTS_URL, status=200, payload=response_districts, repeat=True)
            mocked.get(DISTRICTS_URL_RECOVERED, status=200, payload=response_recovered, repeat=True)
            mocked.get(DISTRICTS_URL_NEW_CASES, status=200, payload=response_new_cases, repeat=True)
            mocked.get(DISTRICTS_URL_NEW_RECOVERED, status=200, payload=response_new_recovered, repeat=True)
            mocked.get(DISTRICTS_URL_NEW_DEATHS, status=200, payload=response_new_death, repeat=True)
            mocked.get(VACCINATIONS_URL, status=200, body=vaccinations_by_state, repeat=True)
            mocked.get(HOSPITALIZATION_URL, status=200, body=hospitalization_numbers, repeat=True)

            parser = RkiCovidParser(session)
            await parser.load_data()
            await parser.load_data()

            assert len(parser.districts) == 412
            for district in parser.districts:
                _validate_district(parser.districts[district])

            for state in parser.states:
                _validate_state(parser.states[state])

            _validate_district_numbers(parser.districts)
            _validate_state_numbers(parser.states)
            _validate_country_numbers(parser.country)
            _validate_vaccination_numbers(parser.states)

def _validate_vaccination_numbers(state: Dict[str, State]) -> None:
    """Validate vaccination numbers against fake fixtures."""
    assert state['Bayern'].vaccinationTotal == 15751380
    assert state['Bayern'].vaccinationFirst == 8277269
    assert state['Bayern'].vaccinationFull == 7868655

def _validate_hospitalization_numbers(state: Dict[str, State]) -> None:
    """Validate hospitalization numbers against fake fixtures."""
    assert state['Bayern'].hospitalizationIncidenceBaby == 4.03
    assert state['Bayern'].hospitalizationIncidenceMerged == 9.49
    assert state['Bayern'].hospitalizationIncidenceChildren == 1.02
    assert state['Bayern'].hospitalizationIncidenceTeen == 2.32
    assert state['Bayern'].hospitalizationIncidenceGrown == 5.11
    assert state['Bayern'].hospitalizationIncidenceSenior == 15.3
    assert state['Bayern'].hospitalizationIncidenceOld == 55.3

    assert state['Bayern'].hospitalizationCasesBaby == 26
    assert state['Bayern'].hospitalizationCasesMerged == 1247
    assert state['Bayern'].hospitalizationCasesChildren == 12
    assert state['Bayern'].hospitalizationCasesTeen == 72
    assert state['Bayern'].hospitalizationCasesGrown == 235
    assert state['Bayern'].hospitalizationCasesSenior == 420
    assert state['Bayern'].hospitalizationCasesOld == 482

def _validate_district_numbers(districts: Dict[int, District]):
    """Validate district numbers against faked fixtures."""
    assert districts['09162'].id == '09162'
    assert districts['09162'].name == 'München'
    assert districts['09162'].county == 'SK München'
    assert districts['09162'].state == 'Bayern'
    assert districts['09162'].population == 1488202
    assert districts['09162'].cases == 79945
    assert districts['09162'].deaths == 1283
    assert districts['09162'].casesPerWeek == 863
    assert districts['09162'].deathsPerWeek == 2
    assert districts['09162'].recovered == 210891
    assert districts['09162'].weekIncidence == 58.0
    assert districts['09162'].casesPer100k == 5371.92
    assert districts['09162'].newCases == 1632
    assert districts['09162'].newDeaths == 1             #
    assert districts['09162'].newRecovered == 210891
    assert districts['09162'].lastUpdate == '09.09.2021, 00:00 Uhr'

def _validate_state_numbers(state: Dict[int, State]) -> None:
    """Validate state numbers against faked fixtures."""
    assert state['Bayern'].name == 'Bayern'
    assert state['Bayern'].population == 13140183
    assert state['Bayern'].cases == 691474
    assert state['Bayern'].deaths == 15449
    assert state['Bayern'].casesPerWeek == 10514
    assert state['Bayern'].deathsPerWeek == 8
    assert state['Bayern'].recovered == 1819168
    assert state['Bayern'].weekIncidence == 80.0
    assert state['Bayern'].casesPer100k == 5262.29
    assert state['Bayern'].newCases == 20958
    assert state['Bayern'].newDeaths == 7
    assert state['Bayern'].newRecovered == 1819168
    assert state['Bayern'].lastUpdate == '09.09.2021, 00:00 Uhr'

def _validate_country_numbers(country: Country) -> None:
    """Validate country numbers against faked fixtures."""
    assert country.name == "Deutschland"
    assert country.population == 83148406
    assert country.cases == 4046112
    assert country.deaths == 92498
    assert country.casesPerWeek == 69457
    assert country.deathsPerWeek == 39
    assert country.recovered == 9829251
    assert country.weekIncidence == 83.5
    assert country.casesPer100k == 4866.13
    assert country.newCases == 73867
    assert country.newDeaths == 22
    assert country.newRecovered == 9829251
    assert country.lastUpdate == '09.09.2021, 00:00 Uhr'

def _validate_district(district: District):
    """Validate all properties are filled."""
    assert district.id is not None
    assert district.name is not None
    assert district.county is not None
    assert district.state is not None
    assert district.population is not None
    assert district.cases is not None
    assert district.deaths is not None
    assert district.casesPerWeek is not None
    assert district.deathsPerWeek is not None
    assert district.recovered is not None
    assert district.weekIncidence is not None
    assert district.casesPer100k is not None
    assert district.newCases is not None
    assert district.newDeaths is not None
    assert district.newRecovered is not None
    assert district.lastUpdate is not None

def _validate_state(state: State):
    """Validate all properties are filled."""
    assert state.cases is not None


src/rki_covid_parser/const.py
"""Constants for the RKI Covid API."""

DISTRICTS_URL = "https://services7.arcgis.com/mOBPykOjAyBO2ZKk/arcgis/rest/services/RKI_Landkreisdaten/FeatureServer/0/query?where=1%3D1&outFields=RS,GEN,EWZ,cases,deaths,county,last_update,cases7_lk,death7_lk,BL&returnGeometry=false&outSR=4326&f=json"
DISTRICTS_URL_RECOVERED = 'https://services7.arcgis.com/mOBPykOjAyBO2ZKk/arcgis/rest/services/Covid19_hubv/FeatureServer/0/query?where=NeuGenesen IN(1,0)&objectIds=&time=&resultType=standard&outFields=AnzahlGenesen,MeldeDatum,IdLandkreis&returnIdsOnly=false&returnUniqueIdsOnly=false&returnCountOnly=false&returnDistinctValues=false&cacheHint=false&orderByFields=IdLandkreis&groupByFieldsForStatistics=IdLandkreis&outStatistics=[{"statisticType":"sum","onStatisticField":"AnzahlGenesen","outStatisticFieldName":"recovered"},{"statisticType":"sum","onStatisticField":"AnzahlFall","outStatisticFieldName":"newCases"},{"statisticType":"sum","onStatisticField":"AnzahlTodesfall","outStatisticFieldName":"newDeaths"},{"statisticType":"max","onStatisticField":"MeldeDatum","outStatisticFieldName":"date"}]&having&resultOffset&resultRecordCount&sqlFormat=none&f=json&token'
DISTRICTS_URL_NEW_CASES = 'https://services7.arcgis.com/mOBPykOjAyBO2ZKk/arcgis/rest/services/Covid19_hubv/FeatureServer/0/query?where=NeuerFall IN(1,-1)&objectIds=&time=&resultType=standard&outFields=AnzahlGenesen,MeldeDatum,IdLandkreis&returnIdsOnly=false&returnUniqueIdsOnly=false&returnCountOnly=false&returnDistinctValues=false&cacheHint=false&orderByFields=IdLandkreis&groupByFieldsForStatistics=IdLandkreis&outStatistics=[{"statisticType":"sum","onStatisticField":"AnzahlGenesen","outStatisticFieldName":"recovered"},{"statisticType":"sum","onStatisticField":"AnzahlFall","outStatisticFieldName":"newCases"},{"statisticType":"sum","onStatisticField":"AnzahlTodesfall","outStatisticFieldName":"newDeaths"},{"statisticType":"max","onStatisticField":"MeldeDatum","outStatisticFieldName":"date"}]&having&resultOffset&resultRecordCount&sqlFormat=none&f=json&token'
DISTRICTS_URL_NEW_RECOVERED = 'https://services7.arcgis.com/mOBPykOjAyBO2ZKk/arcgis/rest/services/Covid19_hubv/FeatureServer/0/query?where=NeuGenesen IN(1,-1)&objectIds=&time=&resultType=standard&outFields=AnzahlGenesen,MeldeDatum,IdLandkreis&returnIdsOnly=false&returnUniqueIdsOnly=false&returnCountOnly=false&returnDistinctValues=false&cacheHint=false&orderByFields=IdLandkreis&groupByFieldsForStatistics=IdLandkreis&outStatistics=[{"statisticType":"sum","onStatisticField":"AnzahlGenesen","outStatisticFieldName":"recovered"},{"statisticType":"sum","onStatisticField":"AnzahlFall","outStatisticFieldName":"newCases"},{"statisticType":"sum","onStatisticField":"AnzahlTodesfall","outStatisticFieldName":"newDeaths"},{"statisticType":"max","onStatisticField":"MeldeDatum","outStatisticFieldName":"date"}]&having&resultOffset&resultRecordCount&sqlFormat=none&f=json&token'
DISTRICTS_URL_NEW_DEATHS = 'https://services7.arcgis.com/mOBPykOjAyBO2ZKk/arcgis/rest/services/Covid19_hubv/FeatureServer/0/query?where=NeuerTodesfall IN(1,-1)&objectIds=&time=&resultType=standard&outFields=AnzahlGenesen,MeldeDatum,IdLandkreis&returnIdsOnly=false&returnUniqueIdsOnly=false&returnCountOnly=false&returnDistinctValues=false&cacheHint=false&orderByFields=IdLandkreis&groupByFieldsForStatistics=IdLandkreis&outStatistics=[{"statisticType":"sum","onStatisticField":"AnzahlGenesen","outStatisticFieldName":"recovered"},{"statisticType":"sum","onStatisticField":"AnzahlFall","outStatisticFieldName":"newCases"},{"statisticType":"sum","onStatisticField":"AnzahlTodesfall","outStatisticFieldName":"newDeaths"},{"statisticType":"max","onStatisticField":"MeldeDatum","outStatisticFieldName":"date"}]&having&resultOffset&resultRecordCount&sqlFormat=none&f=json&token'
HOSPITALIZATION_URL = "https://raw.githubusercontent.com/robert-koch-institut/COVID-19-Hospitalisierungen_in_Deutschland/master/Aktuell_Deutschland_COVID-19-Hospitalisierungen.csv"
VACCINATIONS_URL = 'https://impfdashboard.de/static/data/germany_vaccinations_by_state.tsv'

src/rki_covid_parser/model/area.py
"""State representation with all properties."""
from dataclasses import dataclass


@dataclass
class Area:
    population: int = 0
    cases: int = 0
    deaths: int = 0
    casesPerWeek: int = 0
    deathsPerWeek: int = 0
    recovered: int = 0
    newCases: int = 0
    newDeaths: int = 0
    newRecovered: int = 0

    @property
    def weekIncidence(self) -> float:
        return round(self.casesPerWeek / self.population * 100000, 1)

    @property
    def casesPer100k(self) -> float:
        return round(self.cases / self.population * 100000, 2)


    def __repr__(self):
        return (
            f"{self.__class__.__name__}"
            f"(cases={self.cases!r}, deaths={self.deaths!r}, recovered={self.recovered!r}, newCases={self.newCases!r}\n)"
        )


    def accumulate(self, other):
        assert isinstance(other, Area)

        self.population += other.population
        self.cases += other.cases
        self.deaths += other.deaths
        self.casesPerWeek += other.casesPerWeek
        self.deathsPerWeek += other.deathsPerWeek
        self.recovered += other.recovered
        self.newCases += other.newCases
        self.newDeaths += other.newDeaths
        self.newRecovered += other.newRecovered


src/rki_covid_parser/model/state.py
"""State representation with all properties."""
from dataclasses import dataclass
from .area import Area


@dataclass
class State(Area):
    id: str = None
    name: str = None
    vaccinationTotal: int = 0
    vaccinationFirst: int = 0
    vaccinationFull: int = 0
    hospitalizationCasesMerged: float = 0.0
    hospitalizationIncidenceMerged: float = 0.0
    hospitalizationCasesBaby: float = 0.0
    hospitalizationIncidenceBaby: float = 0.123
    hospitalizationCasesChildren: float = 0.0
    hospitalizationIncidenceChildren: float = 0.0
    hospitalizationCasesTeen: float = 0.0
    hospitalizationIncidenceTeen: float = 0.0
    hospitalizationCasesGrown: float = 0.0
    hospitalizationIncidenceGrown: float = 0.0
    hospitalizationCasesSenior: float = 0.0
    hospitalizationIncidenceSenior: float = 0.0
    hospitalizationCasesOld: float = 0.0
    hospitalizationIncidenceOld: float = 0.0
    lastUpdate: str = None

    def __init__(self, name: str):
        self.name = name

    def __repr__(self):
        return (
            f"{self.__class__.__name__}"
            f"(id={self.id!r}, name={self.name!r}, cases={self.cases!r}, deaths={self.deaths!r}, recovered={self.recovered!r}, newCases={self.newCases!r}\n)"
        )


src/rki_covid_parser/parser.py
"""Parse district responses from RKI Covid numbers."""
import json

import csv
import io
import datetime
import aiohttp
from typing import Dict

from rki_covid_parser.const import (
    DISTRICTS_URL, 
    DISTRICTS_URL_RECOVERED, 
    DISTRICTS_URL_NEW_CASES, 
    DISTRICTS_URL_NEW_RECOVERED, 
    DISTRICTS_URL_NEW_DEATHS, 
    VACCINATIONS_URL, 
    HOSPITALIZATION_URL
)
from rki_covid_parser.model.district import District
from rki_covid_parser.model.state import State
from rki_covid_parser.model.country import Country


VaccinationCode2StateMap = {
    "DE-SH": "Schleswig-Holstein",
    "DE-HH": "Hamburg",
    "DE-NI": "Niedersachsen",
    "DE-HB": "Bremen",
    "DE-NW": "Nordrhein-Westfalen",
    "DE-HE": "Hessen",
    "DE-RP": "Rheinland-Pfalz",
    "DE-BW": "Baden-Württemberg",
    "DE-BY": "Bayern",
    "DE-SL": "Saarland",
    "DE-BE": "Berlin",
    "DE-BB": "Brandenburg",
    "DE-MV": "Mecklenburg-Vorpommern",
    "DE-SN": "Sachsen",
    "DE-ST": "Sachsen-Anhalt",
    "DE-TH": "Thüringen",
}


def generator_attributes_from_features(data):
    assert type(data) == dict
    _features = "features"
    _attributes = "attributes"

    if _features in data:
        for feature in data[_features]:
            assert _attributes in feature
            yield feature[_attributes]


class RkiCovidParser:
    def __init__(self, session: aiohttp.ClientSession):
        self.session = session
        self.districts: Dict[int, District] = {}
        self.states: Dict[str, State] = {}
        self.country = Country()


    async def load_data(self) -> None:
        """load all data and merge results."""
        await self._reset_states()
        await self._load_districts()
        await self._load_districts_recovered()
        await self._load_districts_new_cases()
        await self._load_districts_new_deaths()
        await self._load_districts_new_recovered()
        await self._merge_states()
        await self._merge_country()

        await self._load_hospitalization()
        await self._load_vaccinations()

    async def _reset_states(self) -> None:
        """reset previous loaded values."""
        self.states = {}

    async def _load_districts(self) -> None:
        """load and parse districts."""
        data = await self._load_from_argcis(DISTRICTS_URL)
        await self._extract_districts(data)

    async def _load_districts_recovered(self) -> None:
        """Load and parse recovered"""
        data = await self._load_from_argcis(DISTRICTS_URL_RECOVERED)
        await self._extract_districts_recovered(data)

    async def _load_districts_new_cases(self) -> None:
        """load and parse new cases."""
        data = await self._load_from_argcis(DISTRICTS_URL_NEW_CASES)
        await self._extract_districts_new_cases(data)

    async def _load_districts_new_deaths(self) -> None:
        """load and parse new deaths."""
        data = await self._load_from_argcis(DISTRICTS_URL_NEW_DEATHS)
        await self._extract_districts_new_deaths(data)

    async def _load_districts_new_recovered(self) -> None:
        """load new recovered for districts."""
        data = await self._load_from_argcis(DISTRICTS_URL_NEW_RECOVERED)
        await self._extract_districts_new_recovered(data)

    async def _load_hospitalization(self) -> None:
        """load hospitalization numbers."""
        data = await self._load_csv_from_url(HOSPITALIZATION_URL)
        await self._extract_hospitalization(data)

    async def _load_vaccinations(self) -> None:
        """load vaccinations."""
        data = await self._load_from_tsv(VACCINATIONS_URL)
        await self._extract_vaccinations(data)

    async def _extract_districts(self, data: dict) -> None:
        """iterate through 'attributes' to extract districts."""
        for attributes in generator_attributes_from_features(data):
            id = attributes["RS"]
            self.districts[id] = District(attributes)

    async def _extract_districts_recovered(self, data: dict) -> None:
        """iterate through 'attributes' to extract recovered for districts."""
        for attributes in generator_attributes_from_features(data):
            id = str(attributes["IdLandkreis"]).rjust(5, '0')
            recovered = attributes["recovered"]
            if id in self.districts:
                self.districts[id].recovered = recovered

    async def _extract_districts_new_cases(self, data: dict) -> None:
        """iterate through 'attributes' to extract new cases for districts."""
        for attributes in generator_attributes_from_features(data):
            id = str(attributes["IdLandkreis"]).rjust(5, '0')
            newCases = attributes["newCases"]
            if id in self.districts:
                self.districts[id].newCases = newCases

    async def _extract_districts_new_recovered(self, data: dict) -> None:
        """iterate through 'attributes' to extract new cases for districts."""
        for attributes in generator_attributes_from_features(data):
            id = str(attributes["IdLandkreis"]).rjust(5, '0')
            newRecovered = attributes["recovered"]
            if id in self.districts:
                self.districts[id].newRecovered = newRecovered

    async def _extract_districts_new_deaths(self, data: dict) -> None:
        """iterate through 'attributes' to extract new deaths for districts."""
        for attributes in generator_attributes_from_features(data):
            id = str(attributes["IdLandkreis"]).rjust(5, '0')
            newDeaths = attributes["newDeaths"]
            if id in self.districts:
                self.districts[id].newDeaths = newDeaths

    async def _extract_vaccinations(self, data: csv.DictReader) -> None:
        """iterate through rows to extract vaccinations."""
        assert type(data) == csv.DictReader
        _code = "code"
        _vaccinations_total = "vaccinationsTotal"
        _people_first_total = "peopleFirstTotal"
        _people_full_total = "peopleFullTotal"

        for row in data:
            assert _code in row
            assert _vaccinations_total in row
            assert _people_first_total in row
            assert _people_full_total in row

            if row[_code] in VaccinationCode2StateMap:
                state = VaccinationCode2StateMap[row[_code]]
                if state in self.states:
                    self.states[state].vaccinationTotal = int(row[_vaccinations_total])
                    self.states[state].vaccinationFirst = int(row[_people_first_total])
                    self.states[state].vaccinationFull = int(row[_people_full_total])

    async def _extract_hospitalization(self, data: csv.DictReader) -> None:
        """iterate through rows to extract hospitalization."""
        assert type(data) == csv.DictReader

        _date = "Datum"
        _state = "Bundesland"
        _stateid = "Bundesland_Id"
        _age_group = "Altersgruppe"
        _cases_per_week = "7T_Hospitalisierung_Faelle"
        _incidence_per_week = "7T_Hospitalisierung_Inzidenz"

        for row in data:
            assert _date in row
            assert _state in row
            assert _stateid in row
            assert _age_group in row
            assert _cases_per_week in row
            assert _incidence_per_week in row

            dateValue = str(row[_date])
            stateValue = str(row[_state])
            stateIdValue = int(row[_stateid])
            ageGroupValue = str(row[_age_group])
            try:
                casesValue = int(row[_cases_per_week])
            except ValueError:
                casesValue = 0

            try:
                incidenceValue = float(row[_incidence_per_week])
            except ValueError:
                incidenceValue = 0.0

            # skip older entries
            if dateValue != str(datetime.date.today()):
                continue

            # skip unknown states
            if stateValue not in self.states:
                continue

            if ageGroupValue == '00+':
                self.states[stateValue].hospitalizationCasesMerged = casesValue
                self.states[stateValue].hospitalizationIncidenceMerged = incidenceValue

            if ageGroupValue == '00-04':
                self.states[stateValue].hospitalizationCasesBaby = casesValue
                self.states[stateValue].hospitalizationIncidenceBaby = incidenceValue

            if ageGroupValue == '05-14':
                self.states[stateValue].hospitalizationCasesChildren = casesValue
                self.states[stateValue].hospitalizationIncidenceChildren = incidenceValue

            if ageGroupValue == '15-34':
                self.states[stateValue].hospitalizationCasesTeen = casesValue
                self.states[stateValue].hospitalizationIncidenceTeen = incidenceValue

            if ageGroupValue == '35-59':
                self.states[stateValue].hospitalizationCasesGrown = casesValue
                self.states[stateValue].hospitalizationIncidenceGrown = incidenceValue

            if ageGroupValue == '60-79':
                self.states[stateValue].hospitalizationCasesSenior = casesValue
                self.states[stateValue].hospitalizationIncidenceSenior = incidenceValue

            if ageGroupValue == '80+':
                self.states[stateValue].hospitalizationCasesOld = casesValue
                self.states[stateValue].hospitalizationIncidenceOld = incidenceValue
                    

    async def _load_from_argcis(self, url: str) -> str:
        response = await self.session.get(url)
        # parse data manually, due to missing content-type 'application/json'
        body = await response.text()
        return json.loads(body)

    async def _load_from_tsv(self, url: str) -> str:
        response = await self.session.get(url)
        body = await response.text()
        return csv.DictReader(io.StringIO(body), dialect=csv.excel_tab)

    async def _load_csv_from_url(self, url: str) -> str:
        response = await self.session.get(url)
        body = await response.text()
        return csv.DictReader(io.StringIO(body), dialect=csv.excel)

    async def _merge_states(self) -> None:
        """merge all districts grouped by state."""       

        for district in self.districts.values():
            state = self.states.setdefault(district.state, State(district.state))

            state.accumulate(district)
            state.lastUpdate = district.lastUpdate
            
    async def _merge_country(self) -> None:
        """merge all districts to country."""
        self.country = Country()

        for district in self.districts.values():
            self.country.accumulate(district)
            self.country.lastUpdate = district.lastUpdate
